lib/blocktree/database.go

```diff
@@ -0,0 +1,138 @@
+package blocktree
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"io"
+	"math/big"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+)
+
+// Store stores the blocktree in the underlying db
+func (bt *BlockTree) Store() error {
+	if bt.db == nil {
+		return errors.New("blocktree db is nil")
+	}
+
+	enc, err := bt.Encode()
+	if err != nil {
+		return err
+	}
+
+	return bt.db.Put(common.BlockTreeKey, enc)
+}
+
+// Load loads the blocktree from the underlying db
+func (bt *BlockTree) Load() error {
+	if bt.db == nil {
+		return errors.New("blocktree db is nil")
+	}
+
+	enc, err := bt.db.Get(common.BlockTreeKey)
+	if err != nil {
+		return err
+	}
+
+	return bt.Decode(enc)
+}
+
+// Encode recursively encodes the block tree
+// enc(node) = [32B block hash + 8B arrival time + 8B num children n] | enc(children[0]) | ... | enc(children[n-1])
+func (bt *BlockTree) Encode() ([]byte, error) {
+	return encode(bt.head, []byte{})
+}
+
+func encode(n *node, enc []byte) ([]byte, error) {
+	if n == nil {
+		return enc, nil
+	}
+
+	// encode hash and arrival time
+	enc = append(enc, n.hash[:]...)
+	buf := make([]byte, 8)
+	binary.LittleEndian.PutUint64(buf, n.arrivalTime)
+	enc = append(enc, buf...)
+
+	binary.LittleEndian.PutUint64(buf, uint64(len(n.children)))
+	enc = append(enc, buf...)
+
+	var err error
+	for _, child := range n.children {
+		enc, err = encode(child, enc)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return enc, nil
+}
+
+// Decode recursively decodes an encoded block tree
+func (bt *BlockTree) Decode(in []byte) error {
+	r := &bytes.Buffer{}
+	_, err := r.Write(in)
+	if err != nil {
+		return err
+	}
+
+	hash, err := common.ReadHash(r)
+	if err != nil {
+		return err
+	}
+	arrivalTime, err := common.ReadUint64(r)
+	if err != nil {
+		return err
+	}
+	numChildren, err := common.ReadUint64(r)
+	if err != nil {
+		return err
+	}
+
+	bt.head = &node{
+		hash:        hash,
+		parent:      nil,
+		children:    make([]*node, numChildren),
+		depth:       big.NewInt(0),
+		arrivalTime: arrivalTime,
+	}
+
+	bt.leaves = leafMap{bt.head.hash: bt.head}
+
+	return bt.decode(r, bt.head)
+}
+
+func (bt *BlockTree) decode(r io.Reader, parent *node) error {
```

Comment:
do we really have to have two methods with almost the same name ? decode and Decode ?

Reply:
I'm fine with it, I think it's clear that one is used internally and one is used externally. what would you suggest?

Reply:
I am ok with it for now unless someone has a better idea but a comment here is needed either way

Reply:
well, they have different arguments, the internal one could be called decodeToParent since it has a parent pointer as argument. 
The same thing applies for  `(bt *BlockTree) Encode()` and `func encode`

Reply:
okay, I'll rename it `decodeRecursive` and `encodeRecursive`